### PR TITLE
Implement Replication Stream flow control

### PR DESCRIPTION
**Pull request by xwduan**
- initial commit
- Initial implementation
- Fix

## What changed?

Add implementation for Replication stream flow control
## Why?

To allow receiver back pressure to sender
## How did you test it?

integration test
## Potential risks

n/a
## Documentation

n/a
## Is hotfix candidate?

no

### CODE REVIEW

File: service/history/replication/sender_flow_controller.go
Comment (xwduan) [inline service/history/replication/sender_flow_controller.go:35]: This is an arbitrary number and we can use DC to set that.

File: service/history/replication/receiver_flow_controller.go
Comment (xwduan) [inline service/history/replication/receiver_flow_controller.go:31]: This is an arbitrary number and we can use DC to set that.

File: service/history/replication/sender_flow_controller.go
Comment (xwduan) [inline service/history/replication/sender_flow_controller.go:73]: Initial implementation only support Pause and Resume from receiver side. So it only set to 0 or default number. We can add more granular control in the future, i.e. use error rate to adjust rps.

File: service/history/replication/sender_flow_controller.go
Comment (yux0) [inline service/history/replication/sender_flow_controller.go:82]: can you make sure the rate limiter wait will be updated after set rps with a new value?

File: service/history/replication/stream_sender.go
Comment (xwduan) [inline service/history/replication/stream_sender.go:363]: This should be moved after `convert`, as `convert` may return nil.

File: service/history/replication/sender_flow_controller.go
Comment (xwduan) [reply service/history/replication/sender_flow_controller.go:82]: Good point. From the definition of `Wait` api on rate limiter, it will wait until a token is available. I will double check the implementation and test the scenario:
`Wait` on RPS 0, and then we update the RPS, see if go routine notified.

File: service/history/replication/stream_sender.go
Comment (yux0) [inline service/history/replication/stream_sender.go:363]: task priority?

Comment (xwduan) [reply service/history/replication/stream_sender.go:363]: yeah, I am planning to merge this PR after the the tiered sender implementation. So here is mostly for demoing how to use the wait. After the tiered sender PR merged, this part need to be fixed.

File: service/history/replication/receiver_flow_controller.go
Comment (hai719) [inline service/history/replication/receiver_flow_controller.go:51]: MaxOutstandingTasks?

File: service/history/replication/sender_flow_controller.go
Comment (hai719) [reply service/history/replication/sender_flow_controller.go:35]: Just curious, what is current replication RPS?

Comment (hai719) [reply service/history/replication/sender_flow_controller.go:35]: Are we going to set different RPS for different priorities?

Comment (xwduan) [reply service/history/replication/sender_flow_controller.go]: Yes. The RPS is different for different priority and can set through DC.

File: service/history/replication/sender_flow_controller.go
Comment (xwduan) [reply service/history/replication/sender_flow_controller.go]: Implemented `Resume` and `Pause` with condition variable. Ratelimiter cannot be used in such way.

File: service/history/replication/stream_sender.go
Comment (hai719) [inline service/history/replication/stream_sender.go:531]: Do we need to reset context timeout after RESUME?

File: service/history/replication/stream_sender_flow_controller.go
Comment (hai719) [inline service/history/replication/stream_sender_flow_controller.go:104]: From the code looks like there will be only one waiter for each priority. However this is not explicit given the mismatch of `waiters` (int) and `Signal` here (should be `Broadcast`?).

File: service/history/replication/stream_sender_flow_controller.go
Comment (yux0) [inline service/history/replication/stream_sender_flow_controller.go:48]: why not do it now?

File: service/history/replication/stream_sender_flow_controller.go
Comment (yux0) [inline service/history/replication/stream_sender_flow_controller.go:129]: could it cause problem if it wait on the rate limiter while holding the lock?

File: service/history/replication/stream_sender_flow_controller.go
Comment (yux0) [inline service/history/replication/stream_sender_flow_controller.go:114]: I am not sure if golang would complain concurrent read.

File: service/history/replication/stream_sender_flow_controller.go
Comment (yux0) [inline service/history/replication/stream_sender_flow_controller.go:104]: Broadcast()?

File: common/dynamicconfig/constants.go
Comment (xwduan) [reply common/dynamicconfig/constants.go:2131]: ACK.

File: service/history/replication/stream_sender.go
Comment (xwduan) [reply service/history/replication/stream_sender.go:531]: Good catch. The context used here is for `GetReplicationTasksIter` call, and entire pagination is sharing the same context. We may need to revisit the pagination code to reset context there.

File: service/history/replication/stream_sender_flow_controller.go
Comment (xwduan) [reply service/history/replication/stream_sender_flow_controller.go:48]: My goal in this version is to achieve a simple rate limiting strategy. Will consider doing it after this version.

File: service/history/replication/stream_sender_flow_controller.go
Comment (xwduan) [reply service/history/replication/stream_sender_flow_controller.go:104]: Right, let's use Broadcast here to avoid future regression if multiple thread waiting on this `cond`.

Comment (xwduan) [reply service/history/replication/stream_sender_flow_controller.go:104]: Sure.

File: service/history/replication/stream_sender_flow_controller.go
Comment (xwduan) [reply service/history/replication/stream_sender_flow_controller.go:114]: It should be safe to concurrent read, as long as no write happening. The map here is more like an immutable map, no write happens after initializing.

File: service/history/replication/stream_sender_flow_controller.go
Comment (xwduan) [reply service/history/replication/stream_sender_flow_controller.go:129]: Nice catch. It would, it will block the sender receiver event loop. Let me fix it.

File: service/history/replication/stream_sender.go
Comment (xwduan) [reply service/history/replication/stream_sender.go:531]: Discussed offline, the context fix will be in next PR, which will change the signature of GetReplicationTasksIter.

**Review by yux0 — APPROVED** [review]

**Pull request merged**
